src/pb_tensor.cc

```diff
@@ -391,6 +375,22 @@ PbTensor::FromDLPack(const std::string& name, const py::object& tensor)
     auto ptr_to_tensor = FromDLPackCapsule(
         name, tensor.attr("__dlpack__")(py::arg("stream") = py::int_(1)));
 
+    // In case there is a pending job on the data, where this capsule
+    // is pointing to, we need to wait for it to finish before returning
+    // capsule.
+    // We synchronize on the default stream explicitly since that what we 
+    // pass to external tensor's `__dlpack__` method and in case when memory
+    // is allocated on GPU, PbTensor should be on the default stream.
+    // If external tensor's `__dlpack__` method for some reason does not make
+    // default stream wait for when the capsule is ready, this sync will help. 
+    err = cudaStreamSynchronize(0);
```

Comment:
I don't think this solves the issue with PyTorch. PyTorch doesn't enqueue any work on the default stream. It incorrectly creates another stream instead of default stream and enqueues work on that stream. To fix that problem, we need to create our own stream (or use per-thread default stream) and pass that via dlpack instead of legacy default stream. Feel free to correct me if I'm wrong here.

Reply:
>PyTorch doesn't enqueue any work on the default stream.
What work you are mentioning? If all work was on default stream, then pytorch will schedule it on the default stream. Problem comes from `__dlpack__` implementation, which can't retrieve default stream at the moment. Then it is trying to make that stream wait. That was causing issues, since before this implementation wasn't issuing any work on GPU and counting on external's implementation to make default stream wait. Instead, no work was scheduled and we were risking to return garbadge data. Since any work scheduled on the default stream will also make all non-blocking streams to finish their work, we were counting on it. 

Now when we do `cudaStreamSynchronize(0)` on our side, it will make the device wait for all prior non-blocking streams to finish work, since this call was issued on the default stream and then move on. This also shouldn't affect any other's implementations, since this will be called after the default stream get the `event done` call.

> To fix that problem, we need to create our own stream (or use per-thread default stream) and pass that via dlpack instead of legacy default stream.

I think this is an overkill. PyTorch doesn't support PTDS and they are tricky to work with.